Diffusion/Common/Views/PromptTextField.swift

```diff
@@ -0,0 +1,141 @@
+        }
+        return nil
+    }
+    
+    
+    private var textColor: Color {
+        switch tokenCount {
+        case 0...65:
+            return .green
+        case 66...75:
+            return .orange
+        default:
+            return .red
+        }
+    }
+    
+    // macOS initializer
+    init(text: Binding<String>, isPositivePrompt: Bool, model: Binding<String>) {
+         _textBinding = text
+         self.isPositivePrompt = isPositivePrompt
+        _model = model
+    }
+    
+    //iOS initializer
+    init(text: Binding<String>, isPositivePrompt: Bool, model: String) {
+        _textBinding = text
+        self.isPositivePrompt = isPositivePrompt
+        _model = .constant(model)
+    }
+
+    var body: some View {
+        VStack {
+            HStack {
+                Spacer()
+                if !textBinding.isEmpty {
+                    Text("\(tokenCount)")
+                        .foregroundColor(textColor)
+                    Text(" / \(maxTokenCount)")
+                }
+            }
+            .onReceive(Just(textBinding)) { text in
+                updateTokenCount(newText: text)
+            }
+            .font(.caption)
+            #if os(macOS)
+            TextField(isPositivePrompt ? "Positive prompt" : "Negative Prompt", text: $textBinding,
+                      axis: .vertical)
+                .lineLimit(20)
+                .textFieldStyle(.squareBorder)
+                .listRowInsets(EdgeInsets(top: 0, leading: -20, bottom: 0, trailing: 20))
+                .foregroundColor(textColor == .green ? .white : textColor)
+                .frame(minHeight: 30)
+            #else
+            TextField("Prompt", text: $textBinding, axis: .vertical)
+                .lineLimit(20)
+                .listRowInsets(EdgeInsets(top: 0, leading: -20, bottom: 0, trailing: 20))
+                .foregroundColor(textColor == .green ? .white : textColor)
+                .frame(minHeight: 30)
+            #endif
+        }
+        .onChange(of: model) { model in
+            updateTokenCount(newText: textBinding)
+        }
+    }
+
+    private func updateTokenCount(newText: String) {
+        
+        // ensure that the compiled URL exists
+        guard let compiledURL = compiledURL else { return }
+
+        // Initialize the tokenizer only when it's not created yet or the model changes
+        // Check if the model version has changed
+        let modelVersion = $model.wrappedValue
+        if modelVersion != currentModelVersion {
+            do {
+                tokenizer = try BPETokenizer(
+                    mergesAt: compiledURL.appendingPathComponent("merges.txt"),
+                    vocabularyAt: compiledURL.appendingPathComponent("vocab.json")
+                )
+                currentModelVersion = modelVersion
+            } catch {
+                print("Failed to create tokenizer: \(error)")
+                return
+            }
```

Comment:
Can we maybe try to create the tokenizer once, and only update it if the model selection changes? The reason is that it needs to read those files and build all data structures for them.

Perhaps the tokenizer could be part of the `init` instead of the model. This way we could also remove the `filename` and `downloadedURL` getters from here.

Reply:
It does get created once and reused as long as the model hasn't changed.

        if modelVersion != currentModelVersion {
//only  if the model is NEW since we last updated do we reset the Tokenizer
            do {
                tokenizer = try BPETokenizer(
                    mergesAt: compiledURL.appendingPathComponent("merges.txt"),
                    vocabularyAt: compiledURL.appendingPathComponent("vocab.json")
                )
                currentModelVersion = modelVersion
            } catch {
                print("Failed to create tokenizer: \(error)")
                return
            }
        }
// use the tokenizer, if available, to count the tokens. This could be the same already loaded instance from the last run of this function since "tokenizer = try BPETokenizer" stores the tokenizer at the struct level     @State private var tokenizer: BPETokenizer?
        let (tokens, _) = tokenizer?.tokenize(input: newText) ?? ([], [])

It might be better to use your suggestion of passing in the tokenizer though - somewhere outside of the textfield when the model selection changes init a tokenizer for that model. 

In the current form of the project we don't yet have the additions potentially coming with the ModelInfo work. I propose that ModelInfo eventually become more self aware and that a ModelsViewModel be added as an observable object to track things like the available models given the selected variant (filtering), single source of truth on which is the currently selected model, etc. At that time the ModelsViewModel would also be the right place to load the Tokenizer and provide a valid tokenizer for the currently selected model.

Reply:
You are right, let's deal with that at a later stage!